gematria/datasets/exegesis-benchmark.cc

```diff
@@ -0,0 +1,256 @@
+    std::optional<StringRef> HexValue =
+        TestValue.getAsObject()->getString("Hex");
+    if (!HexValue) ExitWithError("Expected basic block to have hex value");
+
+    std::optional<std::vector<uint8_t>> BytesOr =
+        gematria::ParseHexString(HexValue->str());
+
+    if (!BytesOr.has_value()) ExitWithError("Failed to parse hex value");
+
+    std::vector<gematria::DisassembledInstruction> DisInstructions =
+        ExitOnErr(gematria::DisassembleAllInstructions(
+            *MachineDisassembler, State.getInstrInfo(), State.getRegInfo(),
+            State.getSubtargetInfo(), *MachinePrinter, 0, *BytesOr));
+
+    std::vector<MCInst> Instructions;
+    Instructions.reserve(DisInstructions.size());
+
+    for (const auto &DisInstruction : DisInstructions)
+      Instructions.push_back(DisInstruction.mc_inst);
+
+    BenchmarkCode BenchCode;
+    BenchCode.Key.Instructions = std::move(Instructions);
+
+    const llvm::MCRegisterInfo &MRI = State.getRegInfo();
+
+    for (unsigned I = 0;
+         I < MRI.getRegClass(X86::GR64_NOREX2RegClassID).getNumRegs(); ++I) {
+      RegisterValue RegVal;
+      RegVal.Register =
+          MRI.getRegClass(X86::GR64_NOREX2RegClassID).getRegister(I);
+      RegVal.Value = APInt(64, 0x12345600);
+      BenchCode.Key.RegisterInitialValues.push_back(RegVal);
+    }
+
+    const json::Array *MemoryDefinitions =
+        TestValue.getAsObject()->getArray("MemoryDefinitions");
+
+    if (!MemoryDefinitions)
+      ExitWithError("Expected field MemoryDefinitions does not exist.");
+
+    for (const auto &MemoryDefinitionValue : *MemoryDefinitions) {
+      const json::Object *MemoryDefinitionObject =
+          MemoryDefinitionValue.getAsObject();
+
+      if (!MemoryDefinitionObject) {
+        ExitWithError("Malformed memory definition");
+      }
+
+      std::optional<StringRef> MemoryDefinitionName =
+          MemoryDefinitionObject->getString("Name");
+      std::optional<int64_t> MemoryDefinitionSize =
+          MemoryDefinitionObject->getInteger("Size");
+      // Might need to figure out something better for this as the value should
+      // be an arbitrarily big integer.
+      std::optional<int64_t> MemoryDefinitionHexValue =
+          MemoryDefinitionObject->getInteger("Value");
+
+      if (!MemoryDefinitionName.has_value() ||
+          !MemoryDefinitionSize.has_value() ||
+          !MemoryDefinitionHexValue.has_value())
+        ExitWithError("Malformed memory definition");
+
+      MemoryValue MemVal;
+      MemVal.Value = APInt(64, *MemoryDefinitionHexValue);
+      MemVal.Index = 0;  // Update this to support multiple definitions
+      MemVal.SizeBytes = *MemoryDefinitionSize;
+
+      BenchCode.Key.MemoryValues[MemoryDefinitionName->str()] = MemVal;
+    }
+
+    const json::Array *MemoryMappings =
+        TestValue.getAsObject()->getArray("MemoryMappings");
+
+    if (!MemoryMappings) ExitWithError("Malformed memory mapping");
```

Comment:
I'd put more specific error messages, to make debugging easier. E.g. "Memory mappings are missing.".

That said, this might not be a problem - if the block doesn't access memory, it should be OK to skip this section.